lib/awful/screenshot.lua

```diff
@@ -0,0 +1,870 @@
+
+  -- note the use of _private because screen has no setter
+  if ss.screen then
+    if type(ss.screen) == "number" then
+      ss._private.screen = capi.screen[ss.screen] or aw_screen.focused()
+    end
+  else
+    ss._private.screen = aw_screen.focused()
+  end
+
+  ss._private.geometry = ss.screen.geometry
+  ss:filepath_builder()
+  ss._private.surface = gears.surface(ss.screen.content) -- surface has no setter
+
+  return ss
+
+end
+
+-- Internal function executed when a client window screenshot is taken.
+function screenshot_methods.client(ss)
+	--
+  -- note the use of _private becuase client has no setter
+  if not ss.client then
+    ss._private.client = capi.client.focus 
+  end
+
+  ss._private.geometry = ss.client:geometry()
+  ss:filepath_builder()
+  ss._private.surface = gears.surface(ss.client.content) -- surface has no setter
+  return ss
+end
+
+-- Internal function executed when a snipper screenshot tool is launched.
+function screenshot_methods.snipper(ss)
+
+  if type(ss._private.on_success_cb) ~= "function" then
+    ss._private.on_success_cb = default_on_success_cb -- the cb has no setter
+  end
+
+  local mg_callback = mk_mg_callback(ss)
+  capi.mousegrabber.run(mg_callback, "crosshair")
+
+  return true
+
+end
+
+-- Internal function executed when a snip screenshow (a defined geometry) is
+-- taken.
+function screenshot_methods.snip(ss)
+
+  local root_w, root_h
+  local root_intrsct
+  local snip_surf
+
+  root_w, root_h = root.size()
+
+  if not(ss.geometry and
+         type(ss.geometry.x) == "number" and
+         type(ss.geometry.y) == "number" and
+         type(ss.geometry.width) == "number" and
+         type(ss.geometry.height) == "number") then
+
+    -- Default to entire root window. Also geometry has no setter.
+    ss._private.geometry = {x = 0, y = 0, width = root_w, height = root_h}
+
+  end
+
+  root_intrsct = gears.geometry.rectangle.get_intersection(ss.geometry,
+                                                           {x = 0, y = 0,
+                                                            width = root_w,
+                                                            height = root_h})
+
+  snip_surf = crop_shot(root_intrsct.x, root_intrsct.y,
+                          root_intrsct.width, root_intrsct.height)
+
+  ss:filepath_builder()
+  ss._private.surface = gears.surface(snip_surf) -- surface has no setter
+
+  return ss
+
+end
+
+-- Default method is root
+screenshot_methods.default = screenshot_methods.root
+local default_method_name = "root"
+
+-- Module routines
+
+--- Function to initialize the screenshot library.
+--
+-- Currently only sets the screenshot directory, the filename, prefix and the
+-- snipper tool outline color. More initialization can be added as the API
+-- expands.
+-- @staticfct awful.screenshot.set_defaults
+-- @tparam[opt] table args Table passed with the configuration data.
+-- @treturn boolean true or false depending on success
+function module.set_defaults(args)
```

Comment:
I think this should either use raw varibles, a metatable or even nothing at all. Having a single initialization function for a module is something which ended up causing a lot of issues in the past (`awful.rules`, `awful.utils.shell`). For example, if 2  third-party modules generate screenshot for some reasons, they might end up in a situation where either neither initialize the module, then it misbehave or *both* initialize the module, and it also misbehave.

If possible, I would rather force the defaults to be passed to the screenshot builders every time than storing them. That way it's stateless. If users want to wrap it with their own defaults, they are free to do so.

Reply:
OK, the only defaults right now are the directory (e.g. "/home/username/Images"), the prefix to the filename (e.g. "Screenshot-"), and the color used for the rectangle defining the region captured by the snipper tool.

All screenshot functions accept arguments that override the defaults. The purpose of setting the default was just to allow the user to customize what would happen when no argument is passed.

In effect, if we remove this function, we either need to make it an error to not pass arguments (I don't recommend this) or use whatever hard coded defaults are in the file. There currently are hard coded defaults right now. In fact, the library will work without ever calling the `set_defaults` function, i.e.

```
local awful = require('awful')
ss = awful.screenshot.root()
ss:save()
```

should work just fine (assuming it can find a directory, we need to talk about adding the ability to find the preferred directory). So I can take this out, but this basically allows a customization that the user can set up an environment to use custom defaults. This is similar to the `beautiful` library I guess. Of course, as you point out there is only one copy of those defaults and different areas of the configuration can override one another.

It is up to you, but fwiw I would want the ability to set a system default different that whatever is hard coded.